### DISCUSSION TITLE
Linear thermoelasticity in 2D domain

### DISCUSSION THREAD

**Discussion by Arijit-Majumdar** [firedrakeproject/firedrake#3437] [Firedrake support] [answered]
Hi,
I am trying to solve linear thermoelasticity in a 2D beam in Firedrake. For the temperature I am solving $div(k\nabla T) = 0$. The variational equation is as follows
`aT = Constant(k)*dot(grad(T_test),grad(T_trial))*dx`
`LT = Constant(hT)*T_test*ds(2)`
`solve(aT == LT, T_soln, bcs = bcT, solver_parameters={'ksp_type': 'gmres'})`

The temperature profile looks fine to me. After solving for the temperature I am solving the displacement which follows these equations
$\nabla\cdot\sigma = f$
where $\sigma = \lambda tr(\epsilon) + 2\mu\epsilon$ and $\epsilon = 0.5(\nabla u + \nabla u^T) + \alpha T I_d$. Here $\lambda$ and $\mu$ are constants and $I_d$ is 2x2 identity matrix. I first made two functions for $\sigma$ and $\epsilon$
`def epsilon(u, Temp):`
 ` return 0.5*(grad(u) + grad(u).T) + alpha*Temp*Id`

`def sigma(u, Temp):`
`  return lambda_*div(u)*Id + 2*mu*epsilon(u, Temp)`

Using the variational form I get is
`a = inner(sigma(u, T_soln), epsilon(v,T_soln))*dx`
`L = dot(f, v)*dx`
where u and v are the trial and test functions for displacement and T_soln is the temperature from solving the heat equation. But when I try to solve the above variational equation I get the following error
`ArityMismatch: Adding expressions with non-matching form arguments ('v_0',) vs ().`

I am providing link to my colab notebook. Any help is greatly appreciated.
Colab notebook : https://colab.research.google.com/drive/1JmDhteWrtnqQqy4Tq_OxndxmwrHbI2Us?usp=sharing

Thanks

**Comment by dham** [comment, marked as answer]
Your issue is in the the definition of epsilon. The first term contains u while the second term does not. When you pass in an Argument (a test function or trial function) in as `u`, this means the first summand has an argument and the second does not. That causes the error. Mathematically, what you have is invalid because `a` must be linear in both `u` and `v`, and that definition of epsilon is not linear.
